formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/EncodingUtils.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2017-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+@file:OptIn(ExperimentalSerializationApi::class, ExperimentalUnsignedTypes::class)
+
+package kotlinx.serialization.cbor.internal
+
+import kotlinx.serialization.*
+import kotlinx.serialization.cbor.*
+import kotlinx.serialization.descriptors.*
+
+internal const val FALSE = 0xf4
+internal const val TRUE = 0xf5
+internal const val NULL = 0xf6
+internal const val EMPTY_MAP = 0xa0
+
+internal const val NEXT_HALF = 0xf9
+internal const val NEXT_FLOAT = 0xfa
+internal const val NEXT_DOUBLE = 0xfb
+
+internal const val BEGIN_ARRAY = 0x9f
+internal const val BEGIN_MAP = 0xbf
+internal const val BREAK = 0xff
+
+internal const val ADDITIONAL_INFORMATION_INDEFINITE_LENGTH = 0x1f
+
+internal const val HEADER_BYTE_STRING: Int = 0b010_00000
+internal const val HEADER_STRING: Int = 0b011_00000
+internal const val HEADER_NEGATIVE: Byte = 0b001_00000
+internal const val HEADER_ARRAY: Int = 0b100_00000
+internal const val HEADER_MAP: Int = 0b101_00000
+internal const val HEADER_TAG: Int = 0b110_00000
+
+/** Value to represent an indefinite length CBOR item within a "length stack". */
+internal const val LENGTH_STACK_INDEFINITE = -1
+
+internal const val HALF_PRECISION_EXPONENT_BIAS = 15
+internal const val HALF_PRECISION_MAX_EXPONENT = 0x1f
+internal const val HALF_PRECISION_MAX_MANTISSA = 0x3ff
+
+internal const val SINGLE_PRECISION_EXPONENT_BIAS = 127
+internal const val SINGLE_PRECISION_MAX_EXPONENT = 0xFF
+
+internal const val SINGLE_PRECISION_NORMALIZE_BASE = 0.5f
+
+
+@OptIn(ExperimentalSerializationApi::class)
+internal fun SerialDescriptor.isByteString(index: Int): Boolean {
+    return kotlin.runCatching { getElementAnnotations(index).find { it is ByteString } != null }.getOrDefault(false)
```

Comment:
I tried that already before preparing the updated PR. To my surprise, it lead to a slight, but noticeable performance hit according to JMH, so I kept the `runCatching`

Reply:
1. `find` does not throw exceptions, so `runCatching` literally does nothing here.
2. Current benchmark does not write tags (only verifies them, which is incorrect btw — wrote about that separately)
3. Did you have other benchmark data with labels? I've re-measured PR state vs my diff (attached below) and got no significant difference:
```
PR
Benchmark                Mode  Cnt     Score    Error   Units
CborBaseline.fromBytes  thrpt   20  1637.134 ± 17.891  ops/ms
CborBaseline.toBytes    thrpt   20  2242.036 ± 39.286  ops/ms

PR_WITH_PATCH
Benchmark                Mode  Cnt     Score    Error   Units
CborBaseline.fromBytes  thrpt   20  1641.116 ± 43.964  ops/ms
CborBaseline.toBytes    thrpt   20  2126.622 ± 96.746  ops/ms
```

Reply:
Patch link: https://gist.github.com/sandwwraith/1cc3bf4dc1b1658af34be85dcbb37976

Reply:
I'll re-check (don't know If I still have the un-squashed commits in another branch locally to re-trace my steps), maybe I mixed something up.

While `find` does not throw, I seem to remember the `runCatching` (maybe not in that particular line, but somewhere in what is now EncodingUtils.kt) triggered [KT-59081](https://youtrack.jetbrains.com/issue/KT-59081/WASM-Cant-catch-index-out-of-bounds-and-divide-by-0-exceptions), because `getElementAnnotations` can throws an OOB exception. Apparently, that was just sloppiness, if it is working fine now.

Reply:
OOBE from `getElementAnnotations` means that you've made a mistake with indices somewhere, so I would rather not swallow this exception, since it indicates the bug in the encoder/decoder.

Reply:
done in 897ff727491fdd5f1da6d6eec15b36e7d26a4243